zebra-chain/src/sprout/tests/tree.rs

```diff
@@ -0,0 +1,179 @@
+
+        // Test if the note commitments are counted correctly.
+        assert_eq!(incremental_tree.count(), (i + 1) as u64);
+
+        // Test if we can build the tree from a vector of note commitments
+        // instead of appending only one note commitment to the tree.
+        leaves.push(cm);
+        let ad_hoc_tree = tree::NoteCommitmentTree::from(leaves.clone());
+        let ad_hoc_root_hash = ad_hoc_tree.hash();
+        assert_eq!(hex::encode(ad_hoc_root_hash), test_vectors::ROOTS[i]);
+
+        // The root hashes should match.
+        let ad_hoc_root = ad_hoc_tree.root();
+        assert_eq!(<[u8; 32]>::from(ad_hoc_root), ad_hoc_root_hash);
+
+        // Test if the note commitments are counted correctly.
+        assert_eq!(ad_hoc_tree.count(), (i + 1) as u64);
+    }
+}
+
+#[test]
+fn incremental_roots_with_blocks() -> Result<()> {
+    incremental_roots_with_blocks_for_network(Network::Mainnet)?;
+
+    // TODO Get final Sprout roots for Testnet and uncomment the line below.
+
+    // incremental_roots_with_blocks_for_network(Network::Testnet)?;
+
+    Ok(())
+}
+
+fn incremental_roots_with_blocks_for_network(network: Network) -> Result<()> {
+    let (blocks, sprout_roots) = match network {
+        Network::Mainnet => (&*MAINNET_BLOCKS, &*MAINNET_FINAL_SPROUT_ROOTS),
+
+        // TODO Use `TESTNET_BLOCKS` and `TESTNET_FINAL_SPROUT_ROOTS` once they are
+        // available.
+        Network::Testnet => (&*MAINNET_BLOCKS, &*MAINNET_FINAL_SPROUT_ROOTS),
+    };
+
+    // We pick the initial blocks of the Overwinter network upgrade. This is a
+    // reasonable choice since Overwinter still uses the Sprout shielded pool,
+    // and the block heights have a significant meaning in comparison to a
+    // random Sprout block.
+    let height = NetworkUpgrade::Overwinter
+        .activation_height(network)
+        .unwrap()
+        .0;
+
+    // Build an empty note commitment tree
+    let mut note_commitment_tree = tree::NoteCommitmentTree::default();
+
+    // Load the Overwinter activation block.
+    let overwinter_activation_block = Arc::new(
+        blocks
+            .get(&height)
+            .expect("test vector exists")
+            .zcash_deserialize_into::<Block>()
+            .expect("block is structurally valid"),
+    );
+
+    // Add note commitments from the Overwinter activation block to the tree.
+    for transaction in overwinter_activation_block.transactions.iter() {
+        for sprout_note_commitment in transaction.sprout_note_commitments() {
+            note_commitment_tree
+                .append(*sprout_note_commitment)
+                .expect("we should not be able to fill up the tree");
+        }
+    }
+
+    // Check if the root of the tree of the activation block is correct.
+    let overwinter_activation_block_root =
+        tree::Root(**sprout_roots.get(&height).expect("test vector exists"));
+    assert_eq!(
```

Comment:
This is the new test using the newly obtained test vectors for Overwinter. I must be doing something wrong since the hashes don't match here.

Reply:
I think the issue is that there is a single note commitment tree for each pool; they are not recreated at each network upgrade. In other words, the Sprout tree is created at genesis (or the following block?) and not at the Overwinter upgrade, and the test assumes otherwise.

If you get Sprout roots from the first few blocks after Genesis and change the test to use them, it should work. (Sorry, I should've realized this sooner when you asked about how to extract the roots!)

Note sure if it's possible to keep using these blocks in the test, you'd need to whole tree state at Overwinter activation to start from.

Reply:
Makes total sense, I should have thought of this!